File: App_v2/jieshengshequ-app/src/api/publish.ts
import { http } from './client'

// 文件验证接口
export interface VerifyFileRequest {
  file_path: string
}

export interface VerifyFileResponse {
  exists: boolean
  file_size?: number
  download_url?: string
}

// 发布数据类型定义
export interface PublishFileInfo {
  name: string
  size: number
  type?: string
}

export interface PublishResourceRequest {
  title: string
  content: string
  version?: string
  category?: string
  tags?: string[]
  requirements?: string[]
  files?: PublishFileInfo[]
  screenshots?: PublishFileInfo[]
}

export interface PublishPostRequest {
  title: string
  content: string
  tags?: string[]
  images?: PublishFileInfo[]
  code_snippet?: string
}

export interface PublishResponse {
  id: number
  title: string
  status: string
  created_at: string
}

// 发布资源（REST）
export async function publishResource(data: PublishResourceRequest): Promise<PublishResponse> {
  return http.post<PublishResponse>('/publish/resource', data)
}

// 发布帖子（REST）
export async function publishPost(data: PublishPostRequest): Promise<PublishResponse> {
  return http.post<PublishResponse>('/publish/post', data)
}

// 统一：上传资源文件，经由通用存储端点并绑定 package_id
export async function uploadResourceFile(resourceId: number, file: File): Promise<{ file_path: string; download_url: string; file_size: number }> {
  const formData = new FormData()
  // 使用"file"作为字段名
  formData.append('file', file)
  formData.append('package_id', String(resourceId))
  console.log('开始上传资源文件:', file.name, '大小:', file.size, '资源ID:', resourceId)
  
  // 注意：API_BASE已经包含了/api/v1前缀，所以这里不需要再加
  // 直接使用不带前缀的路径
  const url = '/storage/upload';
  console.log('完整上传URL:', url);
  return http.post<{ file_path: string; download_url: string; file_size: number }>(url, formData)
}

// 上传图片
export async function uploadImage(file: File, packageId?: number): Promise<{ file_path: string; download_url: string; file_size: number }> {
  const formData = new FormData()
  // 使用"file"作为字段名
  formData.append('file', file)
  if (packageId) {
    formData.append('package_id', packageId.toString())
  }
  
  console.log('开始上传图片:', file.name, '大小:', file.size, 'packageId:', packageId)
  // 注意：API_BASE已经包含了/api/v1前缀，所以这里不需要再加
  const url = '/storage/upload';
  console.log('完整上传URL:', url);
  const result = await http.post<{ file_path: string; download_url: string; file_size: number }>(url, formData)
  console.log('上传响应:', result)
  return result
}

// 上传帖子图片并绑定到帖子
export async function uploadPostImage(file: File, postId: number): Promise<{ file_path: string; download_url: string; file_size: number }> {
  const formData = new FormData()
  // 使用"file"作为字段名
  formData.append('file', file)
  formData.append('post_id', String(postId))
  console.log('开始上传帖子图片:', file.name, '大小:', file.size, '帖子ID:', postId)
  // 注意：API_BASE已经包含了/api/v1前缀，所以这里不需要再加
  const url = '/storage/upload';
  console.log('完整上传URL:', url);
  return http.post<{ file_path: string; download_url: string; file_size: number }>(url, formData)
} 

// 验证文件是否已成功上传到存储系统
export async function verifyFile(filePath: string): Promise<VerifyFileResponse> {
  console.log('开始验证文件:', filePath)
  const result = await http.post<VerifyFileResponse>('/storage/verify', { file_path: filePath })
  console.log('验证响应:', result)
  return result
}

File: rope-manager-backend/src/api/v1/publish.rs
use actix_web::{web, HttpRequest, HttpResponse};
use serde_json::json;
use crate::models::{
    PublishResourceRequest, PublishPostRequest, CreatePackageRequest, CreatePostRequest,
    PackageStatus, PostStatus,
};
use crate::services::package_service::PackageService;
use crate::services::post_service::PostService;
use crate::repositories::system_repo::SystemRepository;
use crate::require_auth;
use crate::utils::auth_helper::AuthHelper;

pub fn configure_routes(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("/publish")
            .service(
                web::resource("/resource")
                    .route(web::post().to(publish_resource))
            )
            .service(
                web::resource("/post")
                    .route(web::post().to(publish_post))
            )
    );
}

// 发布资源接口 - 对应前端发布页面的资源类型
async fn publish_resource(
    http_req: HttpRequest,
    req: web::Json<PublishResourceRequest>,
    package_service: web::Data<PackageService>,
    system_repo: web::Data<SystemRepository>,
) -> Result<HttpResponse, actix_web::Error> {
    // 验证用户认证
    let user = require_auth!(&http_req);
    
    log::info!("📦 用户 {} 发布资源: {}", user.username, req.title);
    
    // 验证标题和内容
    if req.title.trim().is_empty() || req.content.trim().is_empty() {
        return Ok(HttpResponse::BadRequest().json(json!({
            "code": 400,
            "message": "标题和内容不能为空"
        })));
    }
    
    // 处理分类ID或分类名称
    let category_id = if let Some(category_str) = &req.category {
        // 首先尝试将其解析为ID（数字）
        if let Ok(id) = category_str.parse::<i32>() {
            Some(id)
        } else {
            // 如果不是数字，则按名称查找
            match system_repo.get_categories().await {
                Ok(categories) => {
                    categories.iter()
                        .find(|cat| cat.name == *category_str)
                        .map(|cat| cat.id)
                },
                Err(e) => {
                    log::error!("获取分类失败: {}", e);
                    None
                }
            }
        }
    } else {
        None
    };
    
    // 创建Package记录
    let create_req = CreatePackageRequest {
        name: req.title.clone(),
        author: user.username.clone(),
        version: req.version.clone(),
        description: Some(req.content.clone()),
        category_id,
        file_url: None, // 发布时文件为空，后续通过上传接口填充
        tags: req.tags.clone(),
        is_pinned: Some(false),
        is_featured: Some(false),
        // 新增字段
        screenshots: None, // 发布时截图为空，后续通过上传接口填充
        cover_image: None,
        requirements: req.requirements.clone(),
        included_files: None,
    };
    
    match package_service.create_package(&create_req).await {
        Ok(package) => {
            log::info!("✅ 资源发布成功: package_id={}, title={}", package.id, package.name);
            Ok(HttpResponse::Ok().json(json!({
                "code": 0,
                "message": "资源发布成功，等待审核",
                "data": {
                    "id": package.id,
                    "title": package.name,
                    "status": "pending",
                    "created_at": package.created_at
                }
            })))
        },
        Err(e) => {
            log::error!("❌ 资源发布失败: {}", e);
            Ok(HttpResponse::BadRequest().json(json!({
                "code": 400,
                "message": format!("发布失败: {}", e)
            })))
        }
    }
}

// 发布帖子接口 - 对应前端发布页面的帖子类型
async fn publish_post(
    http_req: HttpRequest,
    req: web::Json<PublishPostRequest>,
    post_service: web::Data<PostService>,
) -> Result<HttpResponse, actix_web::Error> {
    // 验证用户认证
    let user = require_auth!(&http_req);
    
    log::info!("📝 用户 {} 发布帖子: {}", user.username, req.title);
    
    // 验证标题和内容
    if req.title.trim().is_empty() || req.content.trim().is_empty() {
        return Ok(HttpResponse::BadRequest().json(json!({
            "code": 400,
            "message": "标题和内容不能为空"
        })));
    }
    
    // 创建Post记录
    let create_req = CreatePostRequest {
        title: req.title.clone(),
        content: req.content.clone(),
        category_id: None, // 帖子默认无分类，或者可以添加分类逻辑
        tags: req.tags.clone(),
        status: Some(PostStatus::Published), // 帖子发布后直接设为已发布状态
        // 新增字段
        images: req.images.as_ref().map(|imgs| 
            imgs.iter().map(|_| String::new()).collect() // 发布时图片为空，后续通过上传接口填充
        ),
        code_snippet: req.code_snippet.clone(),
    };
    
    match post_service.create_post(create_req, user.id).await {
        Ok(post_id) => {
            log::info!("✅ 帖子发布成功: post_id={}", post_id);
            Ok(HttpResponse::Ok().json(json!({
                "code": 0,
                "message": "帖子发布成功",
                "data": {
                    "id": post_id,
                    "title": req.title.clone(),
                    "status": "published",
                    "created_at": chrono::Utc::now().to_rfc3339()
                }
            })))
        },
        Err(e) => {
            log::error!("❌ 帖子发布失败: {}", e);
            Ok(HttpResponse::BadRequest().json(json!({
                "code": 400,
                "message": format!("发布失败: {}", e)
            })))
        }
    }
} 